HTML API: avoid processing incomplete tokens

dmsnell opened the pull request:
Trac ticket: Core-60122

Depends on #5721

Currently the Tag Processor assumes that an input document is a _full_ HTML document. Because of this, if there's lingering content after the last tag match it will treat that content as plaintext and skip over it. This is fine for the Tag Processor because if there is lingering content that isn't a valid tag then there's nothing for `next_tag()` to match.

However, in order to support a number of feature expansions it is important to recognize that the remaining content _may_ involve partial syntax elements, such as incomplete tags, attributes, or comments.

In this patch we're adding a mode inside the Tag Processor which will flip when we start parsing HTML syntax but the document finishes before the token does. This will provide the ability to:

 - extend the input document
 - avoid misinterpreting syntax as text
 - guess if we have a complete document, know if we have an incomplete document

In the process of building this patch a few fixes were identified and fixed in the Tag Processor, namely in the handling of incomplete syntax elements.

dmsnell committed:
HTML API: Avoid processing incomplete syntax elements.

The HTML Tag Processor is able to know if it starts parsing a syntax element
and reaches the end of the document before it reaches the end of the element.
In these cases, after this patch, the processor will indicate this condition.

For example, when processing `<div><input type="te` there is an incomplete INPUT
element. The processor will fail to find the INPUT, it will pause right after
the DIV, and `paused_at_incomplete_token()` will return `true`.

This patch doesn't change any existing behaviors, but it adds the new method
to report on the final failure condition. It provides a mechanism for later
use to add chunked parsing to the class, wherein it will be possible to process
a document without having the entire document loaded in memory, for example
when processing unbuffered output.

This is also a necessary change for adding the ability to scan every token in
the document. Currently the Tag Processor only exposes tags as tokens, but it
will need to process `#text` nodes, HTML comments, and other markup in order
to enable behaviors in the HTML Processor and in refactors of existing HTML
processing in Core.

dmsnell committed:
Add ticket number

dmsnell committed:
HTML API: Prevent out-of-bounds string access.

Rebuild of 8905faf8fd390eb888a25a8f546a66e2feccd2c5 after #5725.

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>
Co-authored-by: Bernie Reiter <bernhard-reiter@git.wordpress.org>

sirreal commented:
I ran the tests from #5794 against this branch and it skips fewer tests, but all of those tests seem to fail. **This may be a limitation with #5794 and not this PR.** All of the new failures seem to be with test input that add `<script>` or `<title>` to the `<head>`. Were those tags unsupported before this PR? That would explain the problem.

I pulled the test data and the test suite from that branch (no fixes) and ran the tests with these results:

```sh
./vendor/bin/phpunit -c tests/phpunit/tests/html-api/phpunit.xml --filter Html5lib
```

Trunk:

> Tests: 1693, Assertions: 306, Errors: 30, Failures: 2, Skipped: 1357.

This branch:

> Tests: 1693, Assertions: 423, Errors: 30, Failures: 117, Skipped: 1240.

sirreal commented:
I think the problem is that the tag processor always assumes we're in body, but there are tests that add tags to head. So it's a limitation of the way we're using the tests now.

It's not clear to me why those tags were unsupported before this PR but are supported now… was that intentional?

dmsnell commented:
@sirreal those tests need to be setup differently, because the HTML Processor currently only supports the `<body>` context element inside a fragment parser. that is, if it encounters `<html>` it's not the document HTML tag but a parse error and we're free to bail.

it's not that it assumes we're IN BODY but rather the default parameter to `create_fragment()` is `<body>` and any other context node would cause it to abort. we could setup those tests to handle that.

dmsnell committed:
Fix bug found from HTML5Tests for incomplete SCRIPT tag closer

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

dmsnell commented:
With the merge of [1270194](https://github.com/WordPress/wordpress-develop/pull/5725/commits/1270194d4d6729088a7180b0fd64370a94b9f24a) everything is now passing or skipped in the html5lib test suite, thanks again @sirreal.

There were a couple of fixes I incorporated from your work there, and then I updated that branch with some specific test exclusions based on either a design question or a test that tests behaviors we don't currently support (but will probably do so once we support scanning all tokens).

The necessary change to the html5lib tests was to check for `$p->paused_at_incomplete_token()`

ockham reviewed (approved):
LGTM :shipit: 

Thank you for taking the time to walk me through these changes!
As always, the test coverage also really helps with confidence in these changes 🙂

ockham commented:
Committed to Core in https://core.trac.wordpress.org/changeset/57211.
